services/upload/src/main/java/app/coronawarn/server/services/federation/upload/payload/DiagnosisKeyBatchAssembler.java

```diff
@@ -31,14 +44,53 @@
    */
   public List<DiagnosisKeyBatch> assembleDiagnosisKeyBatch(List<DiagnosisKey> diagnosisKeys) {
     if (diagnosisKeys.isEmpty()) {
+      logger.info("Batches not generated: no pending upload diagnosis keys found.");
       return Collections.emptyList();
     }
-    return List.of(DiagnosisKeyBatch.newBuilder()
-        .addAllKeys(diagnosisKeys.stream()
-            .filter(DiagnosisKey::isConsentToFederation)
-            .map(this::convertKey)
-            .collect(Collectors.toList()))
-        .build());
+    if (diagnosisKeys.size() < uploadConfig.getMinBatchKeyCount()) {
+      logger.info("Batches not generated: less then minimum {} pending upload diagnosis keys.",
+          uploadConfig.getMinBatchKeyCount());
+      return Collections.emptyList();
+    }
+
+    return partionIntoBatches(filterAndConvertToUploadStructure(diagnosisKeys));
+  }
+
+  private List<DiagnosisKeyBatch>  partionIntoBatches(
+      List<app.coronawarn.server.common.protocols.external.exposurenotification.DiagnosisKey> keysToUpload) {
+
+    return  partitionListBySize(keysToUpload, uploadConfig.getMaxBatchKeyCount()).stream()
+                              .map(partition -> DiagnosisKeyBatch.newBuilder().addAllKeys(partition).build())
+                              .collect(Collectors.toList());
+  }
+
+  private <T> List<List<T>> partitionListBySize(List<T> keysToUpload, int size) {
+    ListIterator<T> iterator = keysToUpload.listIterator();
+    List<List<T>> partitions = new ArrayList<List<T>>();
+
+    ArrayList<T> currentPartition = new ArrayList<T>(size / 2);
```

Comment:
When implementing this part, i  made an assumption that, on average, batches (partitions) would have a key count in between half and maximum defined size (5000 from EFGS emails, set to 4000 for safety in our app yaml). With that in mind, this would be a preoptimization to reduce the number doubling and internal array copies when filling it up element by element in the next part of the code.

Rethinking about this, it does not hurt to have this but it might not be needed, just because we run the upload job frequently which means probably batch sizes will be smaller than max/2.